File: server/src/routes/agents.js
const express = require('express');
const { broadcastToRoom } = require('../ws/handler');

function createAgentsRouter(queries) {
  const router = express.Router();

  // GET /api/agents - List all agents
  router.get('/', (req, res) => {
    try {
      const agents = queries.listAgents(req.hashKey);
      res.json({ agents });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // POST /api/agents - Upsert agent definition
  router.post('/', (req, res) => {
    try {
      const { agentId, name, task, startMode, targetUrl, scheduleType, scheduleConfig, enabled } = req.body;
      const effectiveStartMode = startMode || (targetUrl ? 'pinned' : 'ai_routed');

      if (!agentId || !name || !task) {
        return res.status(400).json({ error: 'Missing required fields: agentId, name, task' });
      }
      if (effectiveStartMode === 'pinned' && !targetUrl) {
        return res.status(400).json({ error: 'Pinned agents require targetUrl' });
      }

      queries.upsertAgentData(req.hashKey, {
        agentId, name, task,
        startMode: effectiveStartMode,
        targetUrl: targetUrl || '',
        scheduleType: scheduleType || 'interval',
        scheduleConfig: scheduleConfig || '{}',
        enabled: enabled !== false
      });

      const agent = queries.getAgentData(req.hashKey, agentId);
      broadcastToRoom(req.hashKey, { type: 'agent_updated', agent });
      res.json({ success: true, agent });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // DELETE /api/agents/:agentId - Delete agent
  router.delete('/:agentId', (req, res) => {
    try {
      const result = queries.removeAgent(req.hashKey, req.params.agentId);
      if (result.changes === 0) {
        return res.status(404).json({ error: 'Agent not found' });
      }
      broadcastToRoom(req.hashKey, { type: 'agent_deleted', agentId: req.params.agentId });
      res.json({ success: true });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // PATCH /api/agents/:agentId - Toggle agent enabled/disabled
  router.patch('/:agentId', (req, res) => {
    try {
      const { enabled } = req.body;
      if (typeof enabled !== 'boolean') {
        return res.status(400).json({ error: 'Missing required field: enabled (boolean)' });
      }
      const result = queries.toggleAgentEnabled(req.hashKey, req.params.agentId, enabled);
      if (result.changes === 0) {
        return res.status(404).json({ error: 'Agent not found' });
      }
      const agent = queries.getAgentData(req.hashKey, req.params.agentId);
      broadcastToRoom(req.hashKey, { type: 'agent_updated', agent });
      res.json({ success: true, agent });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // POST /api/agents/:agentId/runs - Report run result
  router.post('/:agentId/runs', (req, res) => {
    try {
      const { agentId } = req.params;
      const body = req.body;

      // Upsert agent data if provided
      if (body.name && body.task && body.targetUrl !== undefined) {
        queries.upsertAgentData(req.hashKey, {
          agentId,
          name: body.name,
          task: body.task,
          startMode: body.startMode || (body.targetUrl ? 'pinned' : 'ai_routed'),
          targetUrl: body.targetUrl || '',
          scheduleType: body.scheduleType || 'interval',
          scheduleConfig: body.scheduleConfig || '{}',
          enabled: body.enabled !== false
        });
      }

      // Record the run
      const run = body.run || {};
      queries.recordRun(req.hashKey, agentId, {
        runId: run.runId || 'run_' + Date.now().toString(36),
        startedAt: run.startedAt || new Date().toISOString(),
        completedAt: run.completedAt || new Date().toISOString(),
        status: run.status || 'unknown',
        result: run.result || null,
        error: run.error || null,
        iterations: run.iterations || 0,
        tokensUsed: run.tokensUsed || 0,
        costUsd: run.costUsd || 0,
        durationMs: run.durationMs || 0,
        executionMode: run.executionMode || 'ai_initial',
        costSaved: run.costSaved || 0
      });

      broadcastToRoom(req.hashKey, {
        type: 'run_completed',
        agentId,
        run: run
      });

      res.json({ success: true });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // GET /api/agents/:agentId/stats - Per-agent cost savings stats
  router.get('/:agentId/stats', (req, res) => {
    try {
      const stats = queries.getPerAgentStats(req.hashKey, req.params.agentId);
      res.json(stats);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // GET /api/agents/:agentId/runs - Get run history (paginated)
  router.get('/:agentId/runs', (req, res) => {
    try {
      const limit = Math.min(parseInt(req.query.limit) || 20, 100);
      const offset = parseInt(req.query.offset) || 0;
      const runs = queries.listRuns(req.hashKey, req.params.agentId, limit, offset);
      const total = queries.countRuns(req.hashKey, req.params.agentId);
      res.json({ runs, total, limit, offset });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  // GET /api/agents/stats - Aggregate stats
  router.get('/stats', (req, res) => {
    try {
      const stats = queries.getAgentStats(req.hashKey);
      res.json(stats);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  return router;
}

module.exports = createAgentsRouter;

File: server-py/app/routers/agents.py
import asyncio
import json
import time

from fastapi import APIRouter, Depends, HTTPException, Query

from ..auth import require_hash_key
from ..db import get_pool
from ..models import (
    AgentListResponse,
    AgentResponse,
    AgentUpsert,
    DeleteResponse,
    RunListResponse,
    RunReportPayload,
)

router = APIRouter(prefix="/api/agents", tags=["agents"])

# SSE clients registry: hash_key -> list[asyncio.Queue]
sse_clients: dict[str, list[asyncio.Queue]] = {}


def broadcast_sse(hash_key: str, data: dict) -> None:
    """Push an event to all SSE clients subscribed to this hash_key."""
    clients = sse_clients.get(hash_key)
    if not clients:
        return
    message = json.dumps(data)
    dead = []
    for i, q in enumerate(clients):
        try:
            q.put_nowait(message)
        except asyncio.QueueFull:
            dead.append(i)
    # Clean up dead queues
    for i in reversed(dead):
        clients.pop(i)
    if not clients:
        sse_clients.pop(hash_key, None)


def _row_to_dict(row) -> dict:
    """Convert an asyncpg Record to a JSON-friendly dict."""
    d = dict(row)
    for k, v in d.items():
        if hasattr(v, "isoformat"):
            d[k] = v.isoformat()
    return d


# --- GET /api/agents ---
@router.get("", response_model=AgentListResponse)
async def list_agents(hash_key: str = Depends(require_hash_key)):
    pool = await get_pool()
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            "SELECT * FROM agents WHERE hash_key = $1 ORDER BY created_at DESC",
            hash_key,
        )
    return AgentListResponse(agents=[_row_to_dict(r) for r in rows])


# --- POST /api/agents ---
@router.post("", response_model=AgentResponse)
async def upsert_agent(body: AgentUpsert, hash_key: str = Depends(require_hash_key)):
    effective_start_mode = body.startMode or ("pinned" if body.targetUrl else "ai_routed")

    if not body.agentId or not body.name or not body.task:
        raise HTTPException(
            status_code=400,
            detail="Missing required fields: agentId, name, task",
        )
    if effective_start_mode == "pinned" and not body.targetUrl:
        raise HTTPException(status_code=400, detail="Pinned agents require targetUrl")

    pool = await get_pool()
    async with pool.acquire() as conn:
        await conn.execute(
            """
            INSERT INTO agents (hash_key, agent_id, name, task, start_mode, target_url, schedule_type, schedule_config, enabled)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8::jsonb, $9)
            ON CONFLICT (hash_key, agent_id) DO UPDATE SET
                name = EXCLUDED.name,
                task = EXCLUDED.task,
                start_mode = EXCLUDED.start_mode,
                target_url = EXCLUDED.target_url,
                schedule_type = EXCLUDED.schedule_type,
                schedule_config = EXCLUDED.schedule_config,
                enabled = EXCLUDED.enabled,
                updated_at = NOW()
            """,
            hash_key,
            body.agentId,
            body.name,
            body.task,
            effective_start_mode,
            body.targetUrl or "",
            body.scheduleType,
            body.scheduleConfig or "{}",
            body.enabled,
        )
        row = await conn.fetchrow(
            "SELECT * FROM agents WHERE hash_key = $1 AND agent_id = $2",
            hash_key,
            body.agentId,
        )

    agent_dict = _row_to_dict(row) if row else None
    broadcast_sse(hash_key, {"type": "agent_updated", "agent": agent_dict})
    return AgentResponse(success=True, agent=agent_dict)


# --- DELETE /api/agents/{agentId} ---
@router.delete("/{agentId}", response_model=DeleteResponse)
async def delete_agent(agentId: str, hash_key: str = Depends(require_hash_key)):
    pool = await get_pool()
    async with pool.acquire() as conn:
        result = await conn.execute(
            "DELETE FROM agents WHERE hash_key = $1 AND agent_id = $2",
            hash_key,
            agentId,
        )
    # result is e.g. "DELETE 1" or "DELETE 0"
    if result.endswith("0"):
        raise HTTPException(status_code=404, detail="Agent not found")

    broadcast_sse(hash_key, {"type": "agent_deleted", "agentId": agentId})
    return DeleteResponse(success=True)


# --- POST /api/agents/{agentId}/runs ---
@router.post("/{agentId}/runs")
async def report_run(
    agentId: str, body: RunReportPayload, hash_key: str = Depends(require_hash_key)
):
    pool = await get_pool()
    async with pool.acquire() as conn:
        # Upsert agent data if provided
        if body.name and body.task and body.targetUrl is not None:
            effective_start_mode = body.startMode or ("pinned" if body.targetUrl else "ai_routed")
            await conn.execute(
                """
                INSERT INTO agents (hash_key, agent_id, name, task, start_mode, target_url, schedule_type, schedule_config, enabled)
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8::jsonb, $9)
                ON CONFLICT (hash_key, agent_id) DO UPDATE SET
                    name = EXCLUDED.name,
                    task = EXCLUDED.task,
                    start_mode = EXCLUDED.start_mode,
                    target_url = EXCLUDED.target_url,
                    schedule_type = EXCLUDED.schedule_type,
                    schedule_config = EXCLUDED.schedule_config,
                    enabled = EXCLUDED.enabled,
                    updated_at = NOW()
                """,
                hash_key,
                agentId,
                body.name,
                body.task,
                effective_start_mode,
                body.targetUrl or "",
                body.scheduleType,
                body.scheduleConfig or "{}",
                body.enabled,
            )

        # Record the run
        run = body.run
        run_id = run.runId or f"run_{int(time.time() * 1000):x}"
        from datetime import datetime, timezone

        now_iso = datetime.now(timezone.utc).isoformat()

        await conn.execute(
            """
            INSERT INTO agent_runs
                (hash_key, agent_id, run_id, started_at, completed_at, status,
                 result, error, iterations, tokens_used, cost_usd, duration_ms,
                 execution_mode, cost_saved)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
            """,
            hash_key,
            agentId,
            run_id,
            run.startedAt or now_iso,
            run.completedAt or now_iso,
            run.status,
            run.result,
            run.error,
            run.iterations,
            run.tokensUsed,
            run.costUsd,
            run.durationMs,
            run.executionMode,
            run.costSaved,
        )

    broadcast_sse(hash_key, {"type": "run_completed", "agentId": agentId, "run": run.model_dump()})
    return {"success": True}


# --- GET /api/agents/{agentId}/runs ---
@router.get("/{agentId}/runs", response_model=RunListResponse)
async def list_runs(
    agentId: str,
    limit: int = Query(20, le=100),
    offset: int = Query(0, ge=0),
    hash_key: str = Depends(require_hash_key),
):
    pool = await get_pool()
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            """
            SELECT * FROM agent_runs
            WHERE hash_key = $1 AND agent_id = $2
            ORDER BY completed_at DESC
            LIMIT $3 OFFSET $4
            """,
            hash_key,
            agentId,
            limit,
            offset,
        )
        total_row = await conn.fetchrow(
            "SELECT COUNT(*) AS count FROM agent_runs WHERE hash_key = $1 AND agent_id = $2",
            hash_key,
            agentId,
        )

    return RunListResponse(
        runs=[_row_to_dict(r) for r in rows],
        total=total_row["count"] if total_row else 0,
        limit=limit,
        offset=offset,
    )
